### PR TITLE
New training script and save model as state_dict

### PR DESCRIPTION
**Pull request by nw13slx**
## Description
- Unified, modular model building and initialization architecture
- Added `nequip-benchmark` script for benchmarking and profiling models
- Now the NequIP code can only run on either plain mode, where root/runname has to be unique, uncreated before, or requeue mode, if root/runname already exists, everything will be automatically appended.

## Motivation and Context

Enable forward/backward compatibility, more modular way of initialization

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:

- [x] My code follows the code style of this project and has been formatted using `black`.
- [x] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [ ] All new and existing tests passed.
- [ ] `example.yaml` (and other relevant `configs/`) have been updated with new or changed options.
- [ ] I have updated `CHANGELOG.md`.

### CODE REVIEW

File: nequip/scripts/benchmark.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        per_atom_time = trim_time / n_atom
+        time_unit_per, time_scale_per = select_unit(per_atom_time)
+        print(
+            f"    \\_ this comes out to {per_atom_time/time_scale_per:g} {time_unit_per}/atom/call"
+        )
+        ns_day = (86400.0 / trim_time) * args.timestep * 1e-6
+        #     day in s^   step/s^       ^ 2fs / step      ^ fs / ns
+        print(
+            f"For this system, at a {args.timestep:.2f}fs timestep, this comes out to {ns_day:.2f} ns/day"
+        )
```
Comment (Linux-cpp-lisp) [inline nequip/scripts/benchmark.py:176]: can you confirm if any of this makes sense?

File: nequip/scripts/deploy.py [hunk 2]
```diff
@@ -126,35 +135,16 @@ def main(args=None):
...
-        model = torch.jit.freeze(model)
-        logging.info("Froze TorchScript model")
+        model = _compile_for_deploy(model)
+        logging.info("Compiled & optimized model.")
 
         # load config
-        # TODO: walk module tree if config does not exist to find params?
-        config_str = (args.train_dir / "config_final.yaml").read_text()
+        config_str = (args.train_dir / "config.yaml").read_text()
         config = yaml.load(config_str, Loader=yaml.Loader)
```
Comment (Linux-cpp-lisp) [inline nequip/scripts/deploy.py:148]: should use Config.from_file, no?

File: nequip/scripts/train.py [hunk 3]
```diff
@@ -1,43 +1,69 @@
...
+        "RescaleEnergyEtc",
+    ],
     dataset_statistics_stride=1,
     default_dtype="float32",
     allow_tf32=False,  # TODO: until we understand equivar issues
     verbose="INFO",
     model_debug_mode=False,
     equivariance_test=False,
     grad_anomaly_mode=False,
+    append=False,
```
Comment (Linux-cpp-lisp) [inline nequip/scripts/train.py:40]: why is this default

File: nequip/scripts/train.py [hunk 4]
```diff
@@ -58,24 +84,24 @@ def parse_command_line(args=None):
         help="enable PyTorch autograd anomaly mode to debug NaN gradients. Do not use for production training!",
         action="store_true",
     )
+    parser.add_argument(
+        "--update_config",
+        help="overwrite the original values in the yaml file",
+        type=str,
+        default=None,
+    )
```
Comment (Linux-cpp-lisp) [inline nequip/scripts/train.py:92]: gotta think about sanity of this

File: nequip/scripts/train.py [hunk 5]
```diff
@@ -118,8 +147,8 @@ def fresh_start(config):
 
         trainer = Trainer(model=None, **dict(config))
 
-    output = trainer.output
-    config.update(output.updated_dict())
+    # what is this
+    config.update(trainer.params)
```
Comment (Linux-cpp-lisp) [inline nequip/scripts/train.py:151]: what is this

File: nequip/scripts/train.py [hunk 6]
```diff
@@ -277,9 +190,56 @@ def fresh_start(config):
     trainer.model = final_model
 
     # Train
-    trainer.train()
+    trainer.update_kwargs(config)
```
Comment (Linux-cpp-lisp) [inline nequip/scripts/train.py:193]: remind me why this exists?

File: nequip/scripts/train.py [hunk 7]
```diff
@@ -277,9 +190,56 @@ def fresh_start(config):
...
+    dictionary = load_file(
+        supported_formats=dict(torch=["pt", "pth"]),
+        filename=restart_file,
+        enforced_format="torch",
+    )
+    dictionary.update(update_config)
+
+    # compare dictionary to config
+    # recursive loop, if same type but different value
+    # raise error
```
Comment (Linux-cpp-lisp) [inline nequip/scripts/train.py:211]: TODO

Comment (Linux-cpp-lisp) [reply nequip/scripts/train.py:211]: but after merge, need to merge

File: nequip/scripts/train.py (hunk 5 above)
Comment (nw13slx) [reply nequip/scripts/train.py:151]: to update wandb data

**Pull request merged**